scripts/input.gd

```diff
@@ -74,26 +98,62 @@ func _get_throw(e):
 		throw = ACT.THROW
 	return throw
 
-func _get_direction(e):
-	dir = -1
-	if e.is_action_pressed('ui_up') and not e.is_action_pressed('ui_down'):
+func _get_direction(e, interaction):
+	var dir = DIR.INVALID
+	var check_action = FuncRef.new()
+
+	if not interaction: return
+
+	# configure funcref's target object and method
+	if interaction == "pressed":
+		check_action = PressInputHandler.new()
+	elif interaction == "released":
+		check_action = ReleaseInputHandler.new()
+	elif interaction == "held":
+		check_action = HoldInputHandler.new()
+
+	# check input with set interaction
+	# we could use a bit mask but whatevss
+	if check_action.handle(e, 'ui_up') and not check_action.handle(e, 'ui_down'):
 		dir = DIR.UP
-	elif e.is_action_pressed('ui_down') and not e.is_action_pressed('ui_up'):
+	elif check_action.handle(e, 'ui_down') and not check_action.handle(e, 'ui_up'):
 		dir = DIR.DOWN
-	if e.is_action_pressed('ui_right') and not e.is_action_pressed('ui_left'):
+	if check_action.handle(e, 'ui_right') and not check_action.handle(e, 'ui_left'):
 		dir = DIR.RIGHT
-	elif e.is_action_pressed('ui_left') and not e.is_action_pressed('ui_right'):
+	elif check_action.handle(e, 'ui_left') and not check_action.handle(e, 'ui_right'):
 		dir = DIR.LEFT
-	if e.is_action_pressed('ui_up') and e.is_action_pressed('ui_right') \
-		and not e.is_action_pressed('ui_down') and not e.is_action_pressed('ui_left'):
+	if check_action.handle(e, 'ui_up') and check_action.handle(e, 'ui_right') \
+		and not check_action.handle(e, 'ui_down') and not check_action.handle(e, 'ui_left'):
 		dir = DIR.UP_RIGHT
-	elif e.is_action_pressed('ui_down') and e.is_action_pressed('ui_left') \
-		and not e.is_action_pressed('ui_up') and not e.is_action_pressed('ui_right'):
+	elif check_action.handle(e, 'ui_down') and check_action.handle(e, 'ui_left') \
+		and not check_action.handle(e, 'ui_up') and not check_action.handle(e, 'ui_right'):
 		dir = DIR.DOWN_LEFT
-	if e.is_action_pressed('ui_down') and e.is_action_pressed('ui_right') \
-		and not e.is_action_pressed('ui_up') and not e.is_action_pressed('ui_left'):
+	if check_action.handle(e, 'ui_down') and check_action.handle(e, 'ui_right') \
+		and not check_action.handle(e, 'ui_up') and not check_action.handle(e, 'ui_left'):
 		dir = DIR.DOWN_RIGHT
-	elif e.is_action_pressed('ui_up') and e.is_action_pressed('ui_left') \
-		and not e.is_action_pressed('ui_down') and not e.is_action_pressed('ui_right'):
+	elif check_action.handle(e, 'ui_up') and check_action.handle(e, 'ui_left') \
+		and not check_action.handle(e, 'ui_down') and not check_action.handle(e, 'ui_right'):
 		dir = DIR.UP_LEFT
+
 	return dir
+
+func do_your_weird_thing(event, act):
```

Comment:
Name suggestion: `dispatch_game_events` (or `signals`, `commands`, etc. instead of `events`)

Also:
1. The helper functions below this one should just return the condition value
2. Their names could be better, like `is_quit(event)` or `check_quit(event)`